redis7/redis.go

```diff
@@ -0,0 +1,371 @@
+		}
+		defer func(txMutex *redsync.Mutex, log *logrus.Logger) {
+			s.log.Tracef("Releasing Redis distributed lock (name=%s)", lockName)
+			_, err := txMutex.Unlock()
+			if err != nil {
+				log.Errorf("Unable to release Redis transaction-level write lock: %s", err)
+			}
+		}(txMutex, s.log)
+	}
+
+	// Start transaction, retrieve/create shelf to operate on
+	s.log.Tracef("Starting Redis transaction (TxPipeline)")
+	pl := s.client.TxPipeline()
+
+	// Perform TX action(s)
+	appError := fn(pl)
+
+	// Observe result, commit/rollback
+	if appError == nil {
+		s.log.Trace("Committing Redis transaction (TxPipeline)")
+		cmdErrs, err := pl.Exec(context.TODO())
+		if err != nil {
+			for _, cmdErr := range cmdErrs {
+				s.log.WithError(cmdErr.Err()).Errorf("Redis pipeline command failed: %s", cmdErr.String())
+			}
+			opts.InvokeOnRollback()
+			return stoabs.ErrCommitFailed
+		}
+		opts.InvokeAfterCommit()
+	} else {
+		s.log.WithError(appError).Warn("Rolling back transaction due to application error")
+		pl.Discard()
+		opts.InvokeOnRollback()
+		return appError
+	}
+
+	return nil
+}
+
+func (s *store) checkOpen() error {
+	s.mux.RLock()
+	defer s.mux.RUnlock()
+	if s.client == nil {
+		return stoabs.ErrStoreIsClosed
+	}
+	return nil
+}
+
+type tx struct {
+	reader redis.Cmdable
+	writer redis.Cmdable
+	store  *store
+}
+
+func (t tx) GetShelfWriter(shelfName string) (stoabs.Writer, error) {
+	return t.store.getShelf(shelfName, t.writer, t.reader), nil
+}
+
+func (t tx) GetShelfReader(shelfName string) stoabs.Reader {
+	return t.store.getShelf(shelfName, nil, t.reader)
+}
+
+func (t tx) Store() stoabs.KVStore {
+	return t.store
+}
+
+func (t tx) Unwrap() interface{} {
+	return t.writer
+}
+
+type shelf struct {
+	name   string
+	prefix string
+	reader redis.Cmdable
+	// writer holds the Redis command writer for writing to the database.
+	// It is separate from the reader because writes use a transactional Redis pipeline, which buffers the commands.
+	// Buffering read commands would result in unexpected behaviour where the reads are only executed on transaction commit.
+	writer redis.Cmdable
+	store  *store
+}
+
+func (s shelf) Put(key stoabs.Key, value []byte) error {
+	if err := s.store.checkOpen(); err != nil {
+		return err
+	}
+	return s.writer.Set(context.TODO(), s.toRedisKey(key), value, 0).Err()
+}
+
+func (s shelf) Delete(key stoabs.Key) error {
+	return s.writer.Del(context.TODO(), s.toRedisKey(key)).Err()
+}
+
+func (s shelf) Get(key stoabs.Key) ([]byte, error) {
+	result, err := s.reader.Get(context.TODO(), s.toRedisKey(key)).Result()
+	if err == redis.Nil {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	return []byte(result), nil
+}
+
+func (s shelf) Iterate(callback stoabs.CallerFn, keyType stoabs.Key) error {
+	var cursor uint64
+	var err error
+	var keys []string
+	for {
+		scanCmd := s.reader.Scan(context.TODO(), cursor, s.toRedisKey(stoabs.BytesKey(""))+"*", int64(resultCount))
```

Comment:
How does the `context.TODO()` relate to the Redis client timeout?

Reply:
Introduced per-TX context nonetheless, because it also applies to BBolt to cancel writable TX commits and/or long-running read operations (any read operation in Redis, iterate and range in BBolt).